FatPkg/EnhancedFatDxe/DiskCache.c

```diff
@@ -8,6 +8,229 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 
 #include "Fat.h"
 
+/**
+  Helper function to check if any LBA is dirty in this cache line.
+
+  @param[in]    CacheTag   - CacheTag to check
+
+  @retval       TRUE       - Cache is Dirty
+                FALSE      - Cache is not Dirty
+
+**/
+STATIC
+BOOLEAN
+IsCacheTagDirty (
+  IN CACHE_TAG  *CacheTag
+  )
+{
+  if (CacheTag == NULL) {
+    ASSERT (CacheTag != NULL);
+    return FALSE;
+  }
+
+  return CacheTag->Dirty;
+}
+
+/**
+  Helper function to clear the dirty state of the cache line.
+
+  @param[in]    CacheTag   - CacheTag to clear
+
+**/
+STATIC
+VOID
+ClearCacheTagDirtyState (
+  IN CACHE_TAG  *CacheTag
+  )
+{
+  if (CacheTag == NULL) {
+    ASSERT (CacheTag != NULL);
+    return;
+  }
+
+  ZeroMem (CacheTag->DirtyBlocks, sizeof (CacheTag->DirtyBlocks));
+  CacheTag->Dirty = FALSE;
+}
+
+/**
+  Helper function to set a bit in a dirty block. This is used to
+  track which blocks to later write to disk.
+
+  @param[in]    BitNumber      - Which bit to set in DirtyBlocks
+  @param[in]    DirtyBlocks    - Array of bits
+
+**/
+STATIC
+VOID
+SetBitInDirtyBlock (
+  IN UINTN         BitNumber,
+  IN DIRTY_BLOCKS  *DirtyBlocks
+  )
+{
+  UINTN  BlockIndex;
+  UINTN  BitIndex;
+
+  //
+  // ASSERTs checking BitNumber are DEBUG build only to verify the assumptions in the
+  // Fat.h defines (See Fat.h lines to describe DIRTY_BITS)
+  //
+  ASSERT (BitNumber < DIRTY_BITS);
+
+  BlockIndex               = BitNumber / DIRTY_BITS_PER_BLOCK;
+  BitIndex                 = BitNumber % DIRTY_BITS_PER_BLOCK;
+  DirtyBlocks[BlockIndex] |= LShiftU64 (1ull, BitIndex);
+}
+
+/**
+  Helper function to check if a particular block is marked dirty or not,
+  so that it can be written to the disk if it is dirty.
+
+  @param[in]    BitNumber      - Which bit to check in DirtyBlocks
+  @param[in]    DirtyBlocks    - Array of bits
+
+**/
+STATIC
+BOOLEAN
+CheckBitInDirtyBlock (

```

Comment:
"CheckXXX" is not a good name. How about "IsBlockDirty (CacheTag, BlockNumber)"?

Reply:
I can change the name